### PR TITLE
Fixed #28201 -- Added null chars validator to CharField and TextField

### PR DESCRIPTION
**Pull request by alejandrozf**
Added default null characters validation to CharField and TextField in
order to avoid raising errors on database layer. Added tests for both
fields.

### CODE REVIEW

File: django/core/validators.py [hunk 1]
```diff
@@ -503,3 +503,37 @@ def get_available_image_extensions():
 validate_image_file_extension = FileExtensionValidator(
     allowed_extensions=get_available_image_extensions(),
 )
+
+
+@deconstructible
+class NullCharValidator(object):
```
Comment (jdufresne) [inline django/core/validators.py]: You can drop `(object)` as master only supports Python3.

File: django/db/models/fields/__init__.py [hunk 2]
```diff
@@ -2093,6 +2094,10 @@ def get_internal_type(self):
 class TextField(Field):
     description = _("Text")
 
+    def __init__(self, *args, **kwargs):
+        super(TextField, self).__init__(*args, **kwargs)
```
Comment (jdufresne) [inline django/db/models/fields/__init__.py]: Here too. You can use `super().__init__(*args, **kwargs)` as master only supports Python3.

Comment (alejandrozf) [inline django/db/models/fields/__init__.py]: thanks @jdufresne !

File: django/core/validators.py [hunk 3]
```diff
@@ -503,3 +503,37 @@ def get_available_image_extensions():
 validate_image_file_extension = FileExtensionValidator(
     allowed_extensions=get_available_image_extensions(),
 )
+
+
+@deconstructible
+class NullCharValidator():
```
Comment (jdufresne) [inline django/core/validators.py]: You can drop these parens as well:

```python
class NullCharValidator:
```

File: django/core/validators.py [hunk 4]
```diff
@@ -503,3 +503,37 @@ def get_available_image_extensions():
...
+        if text is not None:
+            self.text = text
+
+    def __call__(self, value):
+        # using 'str' because of specialized fields
+        if str(value).find('\x00') != -1:
+            raise ValidationError(
+                self.message,
+                code=self.code,
+            )
```
Comment (felixxm) [inline django/core/validators.py]: `ValidationError` can be single lined.

File: django/db/models/fields/__init__.py [hunk 5]
```diff
@@ -1029,6 +1029,7 @@ class CharField(Field):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.validators.append(validators.MaxLengthValidator(self.max_length))
+        self.validators.append(validators.NullCharValidator())
```
Comment (claudep) [inline django/db/models/fields/__init__.py]: One `extend()` instead of two `append()`s?

File: django/core/validators.py [hunk 6]
```diff
@@ -503,3 +503,32 @@ def get_available_image_extensions():
...
+        if message is not None:
+            self.message = message
+        if code is not None:
+            self.code = code
+        if text is not None:
+            self.text = text
+
+    def __call__(self, value):
+        # using 'str' because of specialized fields
+        if str(value).find('\x00') != -1:
```
Comment (claudep) [inline django/core/validators.py]: Maybe: `if '\x00' in str(value):`

File: django/core/validators.py [hunk 7]
```diff
@@ -503,3 +503,32 @@ def get_available_image_extensions():
 validate_image_file_extension = FileExtensionValidator(
     allowed_extensions=get_available_image_extensions(),
 )
+
+
+@deconstructible
+class NullCharValidator:
+    """Validate that the string does not contain null characters."""
+    text = ''
```
Comment (jdufresne) [inline django/core/validators.py]: Is `text` used? If unused, you can remove.

**Comment by charettes** [comment]
Wasn't it discussed that the validation should lie at the form level and not at the model one on the mailing list?

**Comment by claudep** [comment]
Simon, you are totally right! I forgot this when reading the patch.

**Comment by alejandrozf** [comment]
Maybe only adding the null char validator to CharField would be enough? 
Because TextField is converted to CharField when Model to Field translation occurs.

File: django/core/validators.py [hunk 8]
```diff
@@ -503,3 +503,28 @@ def get_available_image_extensions():
 validate_image_file_extension = FileExtensionValidator(
     allowed_extensions=get_available_image_extensions(),
 )
+
+
+@deconstructible
+class NullCharValidator:
+    """Validate that the string does not contain null characters."""
+    message = _('Null characters are not allowed.')
+    code = 'invalid_char_present'
```
Comment (timgraham) [inline django/core/validators.py]: I'd use something like `null_character_not_allowed`

File: django/core/validators.py [hunk 9]
```diff
@@ -503,3 +503,28 @@ def get_available_image_extensions():
...
+
+    def __init__(self, message=None, code=None):
+        if message is not None:
+            self.message = message
+        if code is not None:
+            self.code = code
+
+    def __call__(self, value):
+        # using 'str' because of specialized fields
```
Comment (timgraham) [inline django/core/validators.py]: This comment needs clarification. I'm not sure what "specialized fields" means.

File: django/core/validators.py [hunk 10]
```diff
@@ -503,3 +503,28 @@ def get_available_image_extensions():
 validate_image_file_extension = FileExtensionValidator(
     allowed_extensions=get_available_image_extensions(),
 )
+
+
+@deconstructible
+class NullCharValidator:
```
Comment (timgraham) [inline django/core/validators.py]: maybe something like "ProhibitNullCharactersValidator"

File: tests/validators/tests.py [hunk 11]
```diff
@@ -487,3 +491,14 @@ def test_file_extension_equality(self):
...
+    def test_prohibit_null_characters_validator_equality(self):
+        self.assertEqual(
+            ProhibitNullCharactersValidator(message='some message', code='some code'),
+            ProhibitNullCharactersValidator(message='some message', code='some code')
+        )
+
+        self.assertNotEqual(
+            ProhibitNullCharactersValidator(message='message1', code='code1'),
+            ProhibitNullCharactersValidator(message='message2', code='code2')
+        )
```
Comment (adamchainz) [inline tests/validators/tests.py]: you should also add a test with two instances with no message or code

**Comment by chukkwagon** [comment]
Is there an eta on this optimization reaching the 1.11 branch?

**Comment by timgraham** [comment]
It will not be backported to 1.11 per our [supported versions policy](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions).

**Comment by chukkwagon** [comment]
@timgraham thanks for getting back to me. Why is this classified as an optimization? Without explicit handling of this new exception, any char/text field will generate an application error, which feels more like a critical bug.

**Comment by timgraham** [comment]
I find it easy enough to ignore the error if it happens. As discussed on the ticket, you could also consider using psycopg < 2.7 to avoid the problem.

**Comment by TexasChainstoreManager** [comment]
@charettes What was the justification for not including this in the model field, only the form field? All I can find on the mailing list is this thread:

https://groups.google.com/forum/#!msg/django-developers/D1gvXYCezEc/zl_PZ12FCQAJ

Which doesn't seem to conclude that it should only be on the form field.

I'm trying to fix a bug in a REST API due to NUL chars not being validated before getting to postgres, and I'm not using form fields, so in my case it would be good to have it on the model field.

**Comment by charettes** [comment]
@TexasChainstoreManager it looks like it's part of the discussion you linked to

- https://groups.google.com/d/msg/django-developers/D1gvXYCezEc/5WvrxcYKAwAJ
- https://groups.google.com/d/msg/django-developers/D1gvXYCezEc/C7G_uhsnAQAJ

Which seems to at least favor adding the validation at the form level. I believe the idea was that these characters are mostly provided through user input and thus should be handled there. The fact you're experiencing this issue when validating data pushed to you through an API also supports this point.

Are you using DRF for your API or solely the model level validation? I don't think the model validation layer should be considered able to deal with user input directly.

**Comment by TexasChainstoreManager** [comment]
Yes, I am using DRF for my API. For now I'm subclassing DRF's `CharField` class to add the validation. I suppose the correct thing to do is add this validation to DRF's `CharField` class?

**Comment by charettes** [comment]
I believe you're doing it the right way. I think that could be worth opening a PR against https://github.com/encode/django-rest-framework to add it automatically on Django 2.0+"

**Comment by TexasChainstoreManager** [comment]
Will do, thanks for your advice 😄

**Pull request merged**